factory/api/apiResolverFactory.go

```diff
@@ -469,19 +495,123 @@ func createScQueryElement(
 	}
 
 	argsNewSCQueryService := smartContract.ArgsNewSCQueryService{
-		VmContainer:              vmContainer,
-		EconomicsFee:             args.coreComponents.EconomicsData(),
-		BlockChainHook:           vmFactory.BlockChainHookImpl(),
-		BlockChain:               args.dataComponents.Blockchain(),
-		WasmVMChangeLocker:       args.coreComponents.WasmVMChangeLocker(),
-		Bootstrapper:             args.bootstrapper,
-		AllowExternalQueriesChan: args.allowVMQueriesChan,
-		MaxGasLimitPerQuery:      maxGasForVmQueries,
+		VmContainer:                  vmContainer,
+		EconomicsFee:                 args.coreComponents.EconomicsData(),
+		BlockChainHook:               vmFactory.BlockChainHookImpl(),
+		BlockChain:                   apiBlockchain,
+		WasmVMChangeLocker:           args.coreComponents.WasmVMChangeLocker(),
+		Bootstrapper:                 args.bootstrapper,
+		AllowExternalQueriesChan:     args.allowVMQueriesChan,
+		MaxGasLimitPerQuery:          maxGasForVmQueries,
+		HistoryRepository:            args.processComponents.HistoryRepository(),
+		ShardCoordinator:             args.processComponents.ShardCoordinator(),
+		StorageService:               args.dataComponents.StorageService(),
+		Marshaller:                   args.coreComponents.InternalMarshalizer(),
+		ScheduledTxsExecutionHandler: args.processComponents.ScheduledTxsExecutionHandler(),
+		Uint64ByteSliceConverter:     args.coreComponents.Uint64ByteSliceConverter(),
 	}
 
 	return smartContract.NewSCQueryService(argsNewSCQueryService)
 }
 
+func createNewAccountsAdapterApi(args *scQueryElementArgs, chainHandler data.ChainHandler) (state.AccountsAdapterAPI, error) {
```

Comment:
so we will create a new trie for each SC query service instance.
I think we have this, in one form or another, when we create the AccountsDBApi instance. We just need to create more of those instances and pass them to each SC query service element.

Reply:
We need more AccountsDBApi instances as it is easier to manage the inner trie instances that way. AccountsDB is just a wrapper over the trie.